src/DeterministicFactory_v1.sol

```diff
@@ -0,0 +1,94 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity 0.8.23;
+
+// Internal Interfaces
+import {IDeterministicFactory_v1} from
+    "src/interfaces/IDeterministicFactory_v1.sol";
+
+// External Dependencies
+import {Ownable2Step, Ownable} from "@oz/access/Ownable2Step.sol";
+import {Create2} from "@oz/utils/Create2.sol";
+
+/**
+ * @title   Deterministic Deployment Factory
+ *
+ * @notice  This contract is a factory for deploying contracts using the CREATE2 opcode,
+ *             allowing us to deploy some of our contract at the same address of each EVM-
+ *          compatible network.
+ *
+ * @dev     We limit this to a specific address ('allowedDeployer'), as otherwise anyone
+ *          could deploy our contracts at the addresses we use. This can be problematic
+ *          for contracts that are initialized, i.e. where the owner is only set
+ *          after deployment, which would hand them ownership of the contracts. We intend
+ *          to set this 'allowedDeployer' to the deployment account we use, and then
+ *          set it to the zero address afterwards (until the next deployment takes place).
+ *
+ * @custom:security-contact security@inverter.network
+ *                          In case of any concerns or findings, please refer to our Security Policy
+ *                          at security.inverter.network or email us directly!
+ *
+ * @author  Inverter Network
+ */
+contract DeterministicFactory_v1 is IDeterministicFactory_v1, Ownable2Step {
+    //--------------------------------------------------------------------------
+    // Variables
+
+    // The address that is allowed to use the factory.
+    address public allowedDeployer;
+
+    //--------------------------------------------------------------------------
+    // Modifiers
+
+    /// @notice Modifier to restrict access to the allowed address.
+    modifier onlyAllowed() {
+        if (msg.sender != allowedDeployer) {
+            revert DeterministicFactory__NotAllowed();
+        }
+        _;
+    }
+
+    //--------------------------------------------------------------------------
+    // Constructor
+
+    /// @notice Sets the initial owner of the contract
+    constructor(address _initialOwner) Ownable(_initialOwner) {}
+
+    //--------------------------------------------------------------------------
+    // Public Functions
+
+    /// @inheritdoc IDeterministicFactory_v1
+    function setAllowedDeployer(address _allowedDeployer) external onlyOwner {
+        allowedDeployer = _allowedDeployer;
+        emit DeterministicFactory__AllowedDeployerChanged(_allowedDeployer);
+    }
+
+    /// @inheritdoc IDeterministicFactory_v1
+    function deployWithCreate2(bytes32 salt, bytes calldata code)
```

Comment:
> It automatically uses the senders address under the hood. Had to look it up.

It does **not** use the senders address under the hood, that is the whole idea of using CREATE2. The input factory for CREATE2 are the address of the executing entity (which in this case is the factory contract), as well as the bytecode of the created contract (including constructor data) and the salt.

> Where and how do define the salt. Wouldn't it be easiest to just have it in the factory itself?

The salt is defined outside of the factory, like explained in the `USAGE.md` file, that can look like this: 
```solidity
bytes32 salt = keccak256("inverter-deployment-01");
bytes memory implBytecode = vm.getCode("Governor_v1.sol:Governor_v1");
address govImplementation = factory.deployWithCreate2(salt, implBytecode);
```
The using entity (in our case we use it for the protocol) then needs to handle the use of the right salt at the right time (sometimes different salts may be used during deployment, if otherwise same-bytecode-contracts can't be deployed due to address collisions). So it can't be defined by the factory itself. (I already spoke about this with Felix on how we'll do that in the protocol repo)

Hope that clears it up?